src/ray/raylet/scheduling/cluster_task_manager.h

```diff
@@ -20,7 +20,14 @@ namespace raylet {
 /// Work represents all the information needed to make a scheduling decision.
 /// This includes the task, the information we need to communicate to
 /// dispatch/spillback and the callback to trigger it.
-typedef std::tuple<Task, rpc::RequestWorkerLeaseReply *, std::function<void(void)>> Work;
+struct Work {
+  Task task;
+  rpc::RequestWorkerLeaseReply *reply;
+  std::function<void(void)> callback;
+  bool waiting_worker_popped;
```

Comment:
```suggestion
  bool waiting_worker_popped = false;
```

Reply:
Could we make this and dispatch_finished into a single enum instead of having two separate flags? This seems less brittle, since we only have one state to check. I don't think it's valid to have waiting_worker_popped && dispatch_finished anyway.

Reply:
+1 I think we can just do 

```
enum {
    WAITING # default
    CANCELLED # cancel = true
    SCHEDULED # waiting_worker_popped = true
    DISPATCHED # dispatch_finished = true
}
```

Are there conditions that multiple of them co-exist?

Reply:
@rkooo567 Great! Dose the semantics of `SCHEDULED` make confusing？In raylet, `SCHEDULED ` means node selected and resources occupied?

Reply:
Isn't the worker popped only when it is scheduled though? Or does the pop mean "it will be eventually be scheduled"?

Reply:
I think we can make it explicit and call it `WAITING_FOR_WORKER` instead of `SCHEDULED`. That seems pretty clear.

Reply:
@wuisawesome I found we don't need `SPILLED ` because the item of `Work` will be removed after spilling.

Reply:
Just use three status:
```
enum WorkStatus {
  WAITING,
  WAITING_FOR_WORKER,
  CANCELLED,
};
```